parsers/US_PREPA.py

```diff
@@ -0,0 +1,225 @@
+  dataSource = re.search(r"dataSource: (\{.+\}\]\})\}\);", html).group(1)#Extract object with data
+  dataSource = re.sub(r",\s*\}", "}", dataSource) #,} is valid JavaScript but not valid JSON
+  dataSource = json.loads(dataSource)
+  sourceData = dataSource['data'] # The rest of the dataSource object contains unnecessary stuff like chart theme, label, axis names etc.
+  return sourceData
+
+def convert_timestamp(timestamp_string):
+    """
+    Converts timestamp fetched from website into timezone-aware datetime object
+    Arguments:
+    ----------
+    timestamp_string: timestamp in the format 06/01/2020 08:40:00 AM
+    """
+    timestamp_string = re.sub(r"\s+", " ", timestamp_string)#Replace double spaces with one
+
+    print("PARSED TIMESTAMP",arrow.get(timestamp_string, 'MM/DD/YYYY HH:mm:ss A', tzinfo=timezone_name))
+    return arrow.get(timestamp_string, 'MM/DD/YYYY HH:mm:ss A', tzinfo=timezone_name).datetime
+
+def fetch_production(zone_key='US-PR', session=None, target_datetime=None, logger: logging.Logger = logging.getLogger(__name__)):
+  """Requests the last known production mix (in MW) of a given region
+  Arguments:
+  ----------
+  zone_key: used in case a parser is able to fetch multiple countries
+  session: request session passed in order to re-use an existing session
+  target_datetime: the datetime for which we want production data. If not
+    provided, we should default it to now. If past data is not available,
+    raise a NotImplementedError. Beware that the provided target_datetime is
+    UTC. To convert to local timezone, you can use
+    `target_datetime = arrow.get(target_datetime).to('America/New_York')`.
+    Note that `arrow.get(None)` returns UTC now.
+  logger: an instance of a `logging.Logger` that will be passed by the
+    backend. Information logged will be publicly available so that correct
+    execution of the logger can be checked. All Exceptions will automatically
+    be logged, so when something's wrong, simply raise an Exception (with an
+    explicit text). Use `logger.warning` or `logger.info` for information
+    that can useful to check if the parser is working correctly. A default
+    logger is used so that logger output can be seen when coding / debugging.
+  Returns:
+  --------
+  If no data can be fetched, any falsy value (None, [], False) will be
+    ignored by the backend. If there is no data because the source may have
+    changed or is not available, raise an Exception.
+  A dictionary in the form:
+  {
+    'zoneKey': 'FR',
+    'datetime': '2017-01-01T00:00:00Z',
+    'production': {
+        'biomass': 0.0,
+        'coal': 0.0,
+        'gas': 0.0,
+        'hydro': 0.0,
+        'nuclear': null,
+        'oil': 0.0,
+        'solar': 0.0,
+        'wind': 0.0,
+        'geothermal': 0.0,
+        'unknown': 0.0
+    },
+    'storage': {
+        'hydro': -10.0,
+    },
+    'source': 'mysource.com'
+  }
+  """
+
+  global renewable_output
```

Comment:
Hey @q-- I don't think you need to set this as a global variable given what it's used for.

Reply:
To do https://github.com/tmrowco/electricitymap-contrib/pull/2531/files#diff-a83dfd9ee7c1fe0c19c4ca03828e9f5cR183 I had to make it global, otherwise it only updates the value inside the scope of the for loop

And I'm doing the above because in https://github.com/tmrowco/electricitymap-contrib/pull/2531/files#diff-a83dfd9ee7c1fe0c19c4ca03828e9f5cR195 I check if the page with the renewables breakdown was empty (seen it once, only the one empty data item is present) so I can add renewables to the 'unknown' category instead

There might be a more elegant solution, though

Reply:
I can comment out this line and the program runs exactly the same. All we need to do is clean up some of the variable names imo.

Reply:
That's probably because that logic only kicks in when the graph data at https://aeepr.com/es-pr/generacion/Documents/Unidades_renovables.aspx is empty; it's a backup plan to avoid having the parser fail completely when the data on that page is unavailable.

It was required for that when I added it when writing the parser, so I assume it still is today. Leaving it it in now as said page is working correctly right now so I can't test this easily and I don't want to break the parser.